src/main/scala/it/pagopa/interop/purposeprocess/api/impl/PurposeApiServiceImpl.scala

```diff
@@ -245,29 +237,32 @@ final case class PurposeApiServiceImpl(
   ): Route = authorize(ADMIN_ROLE, M2M_ROLE) {
     logger.info("Retrieving Purposes for EService {}, Consumer {} and States {}", eServiceId, consumerId, states)
 
-    def filterPurposeByUserType(purposes: PurposeManagementDependency.Purposes, userId: UUID): Future[Seq[Purpose]] = {
+    def filterPurposeByOrganizationRole(
+      purposes: PurposeManagementDependency.Purposes,
+      organizationId: UUID
+    ): Future[Seq[Purpose]] = {
       // TODO Bad. This would lead to inconsistent page size during pagination.
       //      Can we use filter parameters on management request?
-      Future
-        .traverse(purposes.purposes)(purpose =>
-          for {
-            userType        <- userType(userId, purpose.eserviceId, purpose.consumerId)
-              .map(Some(_))
-              .recover(_ => None)
-            enhancedPurpose <- userType.traverse(enhancePurpose(purpose, _))
-          } yield enhancedPurpose
-        )
-        .map(_.flatten)
+      for {
+        ps <- Future
+          .traverse(purposes.purposes)(purpose =>
+            for {
+              changedBy       <- getChangedBy(organizationId, purpose.eserviceId, purpose.consumerId)
+                .map(Some(_))
+                .recover(_ => None)
+              enhancedPurpose <- changedBy.traverse(enhancePurpose(purpose, _))
+            } yield enhancedPurpose
+          )
+      } yield ps.flatten
```

Comment:
The external for comprehension clutters a lot IMHO, how about a `.map(_.flatten)` on the `traverse` output?

Reply:
here's my proposal
```suggestion
      Future
        .traverse(purposes.purposes)(purpose =>
          for {
            changedBy       <- getChangedBy(organizationId, purpose.eserviceId, purpose.consumerId).attempt
            enhancedPurpose <- changedBy.traverse(enhancePurpose(purpose, _))
          } yield enhancedPurpose.toOption
        )
        .map(_.flatten)
```